integration-tests/deployment/keystone/deploy.go

```diff
@@ -0,0 +1,807 @@
+					shouldAdd := true
+					for _, existingCapId := range params.HashedCapabilityIds {
+						if existingCapId == proposedCapId {
+							shouldAdd = false
+							break
+						}
+					}
+					if shouldAdd {
+						newCapIds = append(newCapIds, proposedCapId)
+					}
+				}
+				params.HashedCapabilityIds = append(params.HashedCapabilityIds, newCapIds...)
+			}
+			nodeIDToParams[n.ID] = params
+		}
+	}
+
+	var uniqueNodeParams []capabilities_registry.CapabilitiesRegistryNodeParams
+	for _, v := range nodeIDToParams {
+		uniqueNodeParams = append(uniqueNodeParams, v)
+	}
+	lggr.Debugw("unique node params to add", "count", len(uniqueNodeParams))
+	tx, err := req.registry.AddNodes(req.chain.DeployerKey, uniqueNodeParams)
+	if err != nil {
+		err = DecodeErr(kcr.CapabilitiesRegistryABI, err)
+		// no typed errors in the abi, so we have to do string matching
+		// try to add all nodes in one go, if that fails, fall back to 1-by-1
+		if strings.Contains(err.Error(), "NodeAlreadyExists") {
+			lggr.Warn("nodes already exist, falling back to 1-by-1")
+			for _, singleNodeParams := range uniqueNodeParams {
+				tx, err = req.registry.AddNodes(req.chain.DeployerKey, []capabilities_registry.CapabilitiesRegistryNodeParams{singleNodeParams})
+				if err != nil {
+					err = DecodeErr(kcr.CapabilitiesRegistryABI, err)
+					if strings.Contains(err.Error(), "NodeAlreadyExists") {
+						lggr.Warnw("node already exists, skipping", "p2pid", singleNodeParams.P2pId)
+						continue
+					}
+					return nil, fmt.Errorf("failed to call AddNode for node with p2pid %v: %w", singleNodeParams.P2pId, err)
+				}
+				// 1-by-1 tx is pending and we need to wait for it to be mined
+				_, err = req.chain.Confirm(tx)
+				if err != nil {
+					return nil, fmt.Errorf("failed to confirm AddNode of p2pid node %v transaction %s: %w", singleNodeParams.P2pId, tx.Hash().String(), err)
+				}
+				lggr.Debugw("registered node", "p2pid", singleNodeParams.P2pId)
+			}
+		} else {
+			// not a NodeAlreadyExists error, so we can't handle it
+			return nil, fmt.Errorf("failed to call AddNodes for bulk add nodes: %w", err)
+		}
+	} else {
+		// the bulk add tx is pending and we need to wait for it to be mined
+		_, err = req.chain.Confirm(tx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to confirm AddNode confirm transaction %s: %w", tx.Hash().String(), err)
+		}
+	}
+	return &registerNodesResponse{
+		nodeIDToParams: nodeIDToParams,
+	}, nil
+}
+
+type registerDonsRequest struct {
+	registry *capabilities_registry.CapabilitiesRegistry
+	chain    deployment.Chain
+
+	nodeIDToParams    map[string]capabilities_registry.CapabilitiesRegistryNodeParams
+	donToCapabilities map[string][]registeredCapability
+	donToOcr2Nodes    map[string][]*ocr2Node
+}
+
+type registerDonsResponse struct {
+	donInfos map[string]capabilities_registry.CapabilitiesRegistryDONInfo
+}
+
+func sortedHash(p2pids [][32]byte) string {
+	sha256Hash := sha256.New()
+	sort.Slice(p2pids, func(i, j int) bool {
```

Comment:
maybe a way to use/extend sorted methods on https://github.com/smartcontractkit/chainlink/blob/5c51b27bb1350645d08afab2e98e56ba84ae5c28/integration-tests/deployment/environment.go#L128?